pkgs/development/interpreters/python/fetch-python-requirements.nix

```diff
@@ -0,0 +1,127 @@
+# fetchPythonRequirements downlaods python packages specified by a list of
+# pip-style python requirements
+# It also requires a maximum date 'maxDate' being specified.
+# The result will be as if `pip download` would have been executed
+# at the point in time specified by maxDate.
+# This is ensured by putting pip behind a local proxy filtering the
+# api responses from pypi.org to only contain files for which the
+# release date is lower than the specified maxDate.
+
+{
+  # this specifies the python version for which the packages should be downloaded
+  python,
+}:
+
+{
+  lib
+, cacert
+, curl
+, stdenv
+, buildPackages
+, python3
+, ...
+}:
+let
+
+  # we use mitmproxy to filter the pypi responses
+  pythonWithMitmproxy = python3.withPackages (ps: [ ps.mitmproxy ps.python-dateutil ]);
+
+  fetchPythonRequirements = {
+    maxDate,        # maximum release date for packages
+    outputHash,     # hash for the fixed output derivation
+    requirements,   # list of strings of specs
+
+    pipFlags ? [],
+  }:
+    let
+      pythonAttrName = lib.stringAsChars (x: if x == "." then "" else x) python.libPrefix;
+
+      # We need to execute pip from the python version for which we want to download packages.
+      # Pip accepts '--python-version', but this works only for wheel packages.
+      pythonPip = buildPackages."${pythonAttrName}".withPackages (ps: [
+        ps.pip
+        ps.setuptools
+        ps.pkgconfig
+      ]);
+
+      # Normalize package names:
+      #   "Example_Package[extra]" -> "example-package"
+      normalizedNames = map (req:
+        lib.toLower (lib.replaceStrings [ "_" ] [ "-" ] (lib.elemAt (lib.splitString "[" req) 0))
+      ) requirements;
+
+      # fixed output derivation containing packages,
+      # each being symlinked from it's normalized name
+      # Example:
+      #   "$out/werkzeug" will point to "$out/Werkzeug-0.14.1-py2.py3-none-any.whl"
+      self = stdenv.mkDerivation {
+
+        name = "multiple-pypi-packages";
+
+        outputHashMode = "recursive";
+        outputHashAlgo = "sha256";
+        inherit outputHash;
+
+        nativeBuildInputs = with python3.pkgs; [ pythonWithMitmproxy pythonPip curl cacert ];
+
+        phases = [ "buildPhase" ];
```

Comment:
IIRC, explicitly setting phases is discouraged. But don't remember the context

cc @jtojnar

Reply:
See also https://github.com/NixOS/nixpkgs/issues/28910. Mainly the issue is that overriding phases drops `fixupPhase` which breaks most setup hooks. Similar results are caused by overriding `buildCommand` (ex. https://github.com/NixOS/nixpkgs/pull/96552).

But it might not be an issue for tools like this (not sure, maybe you want to run at least `patchShebangsAuto` in `fixupPhase`.

Reply:
fixed in 147d9a50ffe51e7fa1b853001c17d23079926ad1